File: soap-apps/golang/soap_translate.go
// soap_translate.go - Cliente SOAP con traducción en Go
package main

import (
	"encoding/json"
	"encoding/xml"
	"fmt"
	"io/ioutil"
	"log"
	"net/http"
	"net/url"
	"strings"
)

type SoapEnvelope struct {
	XMLName xml.Name `xml:"Envelope"`
	Body    SoapBody `xml:"Body"`
}

type SoapBody struct {
	NumberToWordsResponse NumberToWordsResponse `xml:"NumberToWordsResponse"`
}

type NumberToWordsResponse struct {
	NumberToWordsResult string `xml:"NumberToWordsResult"`
}

func callSOAP(numero string) (string, error) {
	soapRequest := `<?xml version="1.0" encoding="utf-8"?>
<soap:Envelope xmlns:soap="http://schemas.xmlsoap.org/soap/envelope/">
  <soap:Body>
    <NumberToWords xmlns="http://www.dataaccess.com/webservicesserver/">
      <ubiNum>` + numero + `</ubiNum>
    </NumberToWords>
  </soap:Body>
</soap:Envelope>`

	req, err := http.NewRequest("POST", "https://www.dataaccess.com/webservicesserver/NumberConversion.wso", strings.NewReader(soapRequest))
	if err != nil {
		return "", err
	}

	req.Header.Set("Content-Type", "text/xml; charset=utf-8")

	client := &http.Client{}
	resp, err := client.Do(req)
	if err != nil {
		return "", err
	}
	defer resp.Body.Close()

	body, err := ioutil.ReadAll(resp.Body)
	if err != nil {
		return "", err
	}

	var envelope SoapEnvelope
	err = xml.Unmarshal(body, &envelope)
	if err != nil {
		return "", err
	}

	return envelope.Body.NumberToWordsResponse.NumberToWordsResult, nil
}

func traducir(texto string) (string, error) {
	url := fmt.Sprintf("https://translate.googleapis.com/translate_a/single?client=gtx&sl=en&tl=es&dt=t&q=%s", url.QueryEscape(texto))
	resp, err := http.Get(url)
	if err != nil {
		return "", err
	}
	defer resp.Body.Close()

	body, _ := ioutil.ReadAll(resp.Body)
	var result []interface{}
	err = json.Unmarshal(body, &result)
	if err != nil {
		return "", err
	}

	if arr, ok := result[0].([]interface{}); ok {
		if first, ok := arr[0].([]interface{}); ok {
			return first[0].(string), nil
		}
	}
	return "", fmt.Errorf("error en traducción")
}

func main() {
	http.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
		numero := r.URL.Query().Get("n")
		if numero == "" {
			numero = "10"
		}

		resultadoIngles, err := callSOAP(numero)
		if err != nil {
			fmt.Fprintf(w, "Error SOAP: %s", err.Error())
			return
		}

		resultadoEspanol, err := traducir(resultadoIngles)
		if err != nil {
			fmt.Fprintf(w, "Error traducción: %s", err.Error())
			return
		}

		fmt.Fprintf(w, "Numero %s en español: %s", numero, resultadoEspanol)
	})

	fmt.Println("Servidor iniciado en http://localhost:8081")
	fmt.Println("Ejemplo: http://localhost:8081/?n=10")
	log.Fatal(http.ListenAndServe(":8081", nil))
}

File: soap-apps/java/SoapTranslate/src/main/java/SoapTranslate.java
import spark.Spark;
import java.net.HttpURLConnection;
import java.net.URL;
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.io.OutputStream;

public class SoapTranslate {
    
    static String callSOAP(String numero) throws Exception {
        String soapRequest = "<?xml version=\"1.0\" encoding=\"utf-8\"?>" +
            "<soap12:Envelope xmlns:soap12=\"http://www.w3.org/2003/05/soap-envelope\">" +
            "<soap12:Body>" +
            "<NumberToWords xmlns=\"http://www.dataaccess.com/webservicesserver/\">" +
            "<ubiNum>" + numero + "</ubiNum>" +
            "</NumberToWords>" +
            "</soap12:Body>" +
            "</soap12:Envelope>";
        
        URL url = new URL("https://www.dataaccess.com/webservicesserver/NumberConversion.wso");
        HttpURLConnection conn = (HttpURLConnection) url.openConnection();
        conn.setRequestMethod("POST");
        conn.setRequestProperty("Content-Type", "application/soap+xml; charset=utf-8");
        conn.setDoOutput(true);
        
        try (OutputStream os = conn.getOutputStream()) {
            os.write(soapRequest.getBytes("UTF-8"));
        }
        
        BufferedReader reader = new BufferedReader(new InputStreamReader(conn.getInputStream()));
        StringBuilder response = new StringBuilder();
        String line;
        while ((line = reader.readLine()) != null) {
            response.append(line);
        }
        reader.close();
        
        String responseStr = response.toString();
        String[] parts = responseStr.split("<[^>]+>");
        for (String part : parts) {
            String trimmed = part.trim();
            if (!trimmed.isEmpty() && trimmed.matches("[a-zA-Z ]+")) {
                return trimmed;
            }
        }
        
        return "Error: " + responseStr;
    }
    
    static String traducir(String texto) throws Exception {
        String translateUrl = "https://translate.googleapis.com/translate_a/single?client=gtx&sl=en&tl=es&dt=t&q=" + 
                     java.net.URLEncoder.encode(texto, "UTF-8");
        HttpURLConnection conn = (HttpURLConnection) new URL(translateUrl).openConnection();
        conn.setRequestProperty("User-Agent", "Mozilla/5.0");
        BufferedReader reader = new BufferedReader(new InputStreamReader(conn.getInputStream()));
        StringBuilder response = new StringBuilder();
        String line;
        while ((line = reader.readLine()) != null) response.append(line);
        reader.close();
        
        String json = response.toString();
        
        // El formato es: [[["traducción","original",...],...],...]
        int start = json.indexOf("[[[\"") + 4;
        int end = json.indexOf("\"", start);
        
        if (start > 3 && end > start) {
            return json.substring(start, end);
        }
        
        return "Error al traducir: " + json;
    }
    
    public static void main(String[] args) throws Exception {
        Spark.port(4568);
        Spark.get("/", (req, res) -> {
            String numero = req.queryParams("n");
            if (numero == null || numero.isEmpty()) numero = "10";
            
            try {
                String resultadoIngles = callSOAP(numero);
                String resultadoEspanol = traducir(resultadoIngles);
                return "Numero " + numero + " en español: " + resultadoEspanol;
            } catch (Exception e) {
                return "Error: " + e.getMessage();
            }
        });
        
        System.out.println("Servidor iniciado en http://localhost:4568");
        System.out.println("Ejemplo: http://localhost:4568/?n=10");
        
        Thread.currentThread().join();
    }
}